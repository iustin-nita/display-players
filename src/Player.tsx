import React from 'react';
import { Card, Icon, Button, Image, Item, Label } from 'semantic-ui-react';
import { sanitizeWeight } from './helpers';

import { PlayerType } from './types';

type PlayerProps = {
  player: PlayerType;
  lastGames: { wins: number; loses: number };
};

export const Player = ({ player, lastGames }: PlayerProps) => {
  return (
    <Item>
      <Item.Image src={player.picture} />
      <Item.Content>
        <Item.Header as="a">
          {`${player.firstname} ${player.lastname}`}
        </Item.Header>
        <Item.Meta>
          <span className="item-meta">
            <Image
              avatar
              circular
              size="mini"
              aria-label="country"
              src={player.country.picture}
            />
            <span className="item-meta-label">{player.country.code}</span>
          </span>
        </Item.Meta>
        <Item.Meta>
          <span className="item-meta">
            <Icon aria-label="rank" name="chart line" />
            <span className="item-meta-label">
              Rank: <strong>{player.data.rank}</strong> -{' '}
              <span className="item-meta-label">
                {player.data.points} Points
              </span>
            </span>
          </span>
        </Item.Meta>
        <Item.Description>
          A description for {`${player.firstname} ${player.lastname}`} which may
          flow for several lines and give context to the content.
        </Item.Description>
        <Item.Extra>
          {player.sex === 'M' ? (
            <span>
              <Icon name="male" fitted aria-label="male" /> Male
            </span>
          ) : (
            <span>
              <Icon name="female" fitted aria-label="female" /> Female
            </span>
          )}
          {/* <span>
              <Flag name={flag} />
              {player.country.code}
            </span> */}
          <span>
            <Icon name="weight" />
            {sanitizeWeight(player.data.weight)}
          </span>
          <span>
            <Icon name="resize vertical" />
            {player.data.height} cm
          </span>
        </Item.Extra>
        <Item.Extra>
          <span>
            Wins: <strong>{lastGames.wins}</strong> Loses:{' '}
            <strong>{lastGames.loses}</strong>
          </span>
        </Item.Extra>
      </Item.Content>
    </Item>
  );
};

export default Player;
