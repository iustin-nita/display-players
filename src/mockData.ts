export const fakePlayer = {
  firstname: 'Stan',
  lastname: 'Wawrinka',
  shortname: 'S.WAW',
  sex: 'M',
  country: {
    picture: 'https://i.eurosport.com/_iss_/geo/country/flag/large/2213.png',
    code: 'SUI'
  },
  picture: 'https://i.eurosport.com/_iss_/person/pp_clubteam/large/325225.jpg',
  data: {
    rank: 21,
    points: 1784,
    weight: 81000,
    height: 183,
    age: 33,
    last: [1, 1, 1, 0, 1]
  }
};

export const fakeLastGames = { wins: 3, loses: 2 };

export const fakeState = {
  data: fakePlayer,
  isLoading: false,
  isError: false
};
