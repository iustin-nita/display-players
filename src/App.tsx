import React, { useReducer, useEffect } from 'react';
import axios from 'axios';
import { Dimmer, Loader, Header, Item, Container } from 'semantic-ui-react';
import Player from './Player';
import { State, Actions, PlayerType } from './types';
import { sanitizeLastGames } from './helpers';

export const API =
  'https://eurosportdigital.github.io/eurosport-web-developer-recruitment/headtohead.json';

export const dataFetchReducer = (state: State, action: Actions): State => {
  switch (action.type) {
    case 'FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'FETCH_ERROR':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    default:
      throw new Error();
  }
};

const initialData: State = {
  data: [],
  isError: false,
  isLoading: false,
};

export const App: React.FC = () => {
  const [{ data, isError, isLoading }, dispatch] = useReducer(
    dataFetchReducer,
    initialData
  );

  useEffect(() => {
    const fetchData = async () => {
      dispatch({ type: 'FETCH_INIT' });
      try {
        const result = await axios(API);
        setTimeout(() => {
          dispatch({ type: 'FETCH_SUCCESS', payload: result.data.players });
        }, 2000);
      } catch (error) {
        dispatch({ type: 'FETCH_ERROR' });
      }
    };

    fetchData();
  }, []);

  const style = {
    h1: {
      margin: '3em 0',
    },
  };

  return (
    <div>
      <Header as="h1" content="Eurosport" style={style.h1} textAlign="center" />

      <Container>
        {isError && (
          <h1 className="error-message">
            Something went wrong ... Try to disable your ADBlock
          </h1>
        )}
        {isLoading ? (
          <Dimmer active>
            <Loader />
          </Dimmer>
        ) : (
          <Item.Group divided>
            {data.map((player: PlayerType) => {
              const lastGames = sanitizeLastGames(player.data.last);
              return <Player player={player} lastGames={lastGames} />;
            })}
          </Item.Group>
        )}
      </Container>
    </div>
  );
};

export default App;
