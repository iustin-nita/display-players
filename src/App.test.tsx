import React from 'react';
import renderer from 'react-test-renderer';
import { Item } from 'semantic-ui-react';
import Enzyme, { configure, shallow, mount } from 'enzyme';
import axios from 'axios';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import App, { dataFetchReducer } from './App';
import Player from './Player';
import { fakePlayer, fakeLastGames, fakeState } from './mockData';
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('App', () => {
  describe('Reducer', () => {
    it('should set isLoading when starting to fetch', () => {
      const state = fakeState;
      const newState = dataFetchReducer(state, {
        type: 'FETCH_INIT',
      });
      expect(newState).toEqual({ ...state, isLoading: true });
    });

    it('should reset the error when starting to fetch', () => {
      const state = { ...fakeState, isError: true };
      const newState = dataFetchReducer(state, {
        type: 'FETCH_INIT',
      });
      expect(newState).toEqual({
        ...state,
        isLoading: true,
        isError: false,
      });
    });
    it('should reset the error/loading if fetch is sucessful', () => {
      const state = { ...fakeState, isError: true };
      const newState = dataFetchReducer(state, {
        type: 'FETCH_SUCCESS',
        payload: [fakePlayer],
      });
      expect(newState).toEqual({
        data: [fakePlayer],
        isLoading: false,
        isError: false,
      });
    });
    it('should set the error when fetch fails', () => {
      const state = fakeState;
      const newState = dataFetchReducer(state, {
        type: 'FETCH_ERROR',
      });
      expect(newState.isError).toBeTruthy();
    });
  });

  it('fetches async data and renders the inner Player Component', async done => {
    const promise = new Promise((resolve, reject) =>
      setTimeout(
        () =>
          resolve({
            data: [fakePlayer],
          }),
        100
      )
    );

    mockedAxios.get = jest.fn(() => promise);

    const wrapper = mount(<App />);
    expect(wrapper.find(Item).length).toEqual(0);
    // await mockedAxios.get.mockResolvedValue({ data: [fakePlayer] });
    promise.then(() => {
      setImmediate(() => {
        wrapper.update();
        expect(wrapper.find(Item).length).toEqual(1);
        mockedAxios.get.mockClear();
        done();
      });
    });
  });

  it('fetches async data but fails', done => {
    const promise = new Promise((resolve, reject) =>
      setTimeout(() => reject(new Error('Nope !')), 100)
    );
    mockedAxios.get = jest.fn(() => promise);
    const wrapper = mount(<App />);
    promise.catch(() => {
      setImmediate(() => {
        wrapper.update();
        expect(wrapper.find(Item).length).toEqual(0);
        expect(wrapper.find('.error-message').length).toEqual(1);
        mockedAxios.get.mockClear();
        done();
      });
    });
  });

  test('snapshot renders', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Player', () => {
  test('snapshot renders', () => {
    const component = renderer.create(
      <Player player={fakePlayer} lastGames={fakeLastGames} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
