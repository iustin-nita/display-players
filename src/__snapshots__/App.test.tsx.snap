// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div>
  <h1
    className="ui center aligned header"
    style={
      Object {
        "margin": "3em 0",
      }
    }
  >
    Eurosport
  </h1>
  <div
    className="ui container"
  >
    <div
      className="ui divided items"
    />
  </div>
</div>
`;

exports[`Player snapshot renders 1`] = `
<div
  className="item"
>
  <div
    className="image"
  >
    <img
      src="https://i.eurosport.com/_iss_/person/pp_clubteam/large/325225.jpg"
    />
  </div>
  <div
    className="content"
  >
    <a
      className="header"
    >
      Stan Wawrinka
    </a>
    <div
      className="meta"
    >
      <span
        className="item-meta"
      >
        <img
          aria-label="country"
          className="ui mini avatar circular image"
          src="https://i.eurosport.com/_iss_/geo/country/flag/large/2213.png"
        />
        <span
          className="item-meta-label"
        >
          SUI
        </span>
      </span>
    </div>
    <div
      className="meta"
    >
      <span
        className="item-meta"
      >
        <i
          aria-label="rank"
          className="chart line icon"
          onClick={[Function]}
        />
        <span
          className="item-meta-label"
        >
          Rank: 
          <strong>
            21
          </strong>
           -
           
          <span
            className="item-meta-label"
          >
            1784
             Points
          </span>
        </span>
      </span>
    </div>
    <div
      className="description"
    >
      A description for 
      Stan Wawrinka
       which may flow for several lines and give context to the content.
    </div>
    <div
      className="extra"
    >
      <span>
        <i
          aria-label="male"
          className="male fitted icon"
          onClick={[Function]}
        />
         Male
      </span>
      <span>
        <i
          aria-hidden="true"
          className="weight icon"
          onClick={[Function]}
        />
        81 kg
      </span>
      <span>
        <i
          aria-hidden="true"
          className="resize vertical icon"
          onClick={[Function]}
        />
        183
         cm
      </span>
    </div>
    <div
      className="extra"
    >
      <span>
        Wins: 
        <strong>
          3
        </strong>
         Loses:
         
        <strong>
          2
        </strong>
      </span>
    </div>
  </div>
</div>
`;
